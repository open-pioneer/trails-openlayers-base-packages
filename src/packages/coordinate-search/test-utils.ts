// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { act, fireEvent } from "@testing-library/react";

export function showDropdown(projSelect: HTMLElement) {
    // open dropdown to include options in snapshot; react-select creates list of options in dom after opening selection
    act(() => {
        fireEvent.keyDown(projSelect, { key: "ArrowDown" });
    });
}

export function getCurrentOptions() {
    return Array.from(
        document.getElementsByClassName("coordinate-input-select__option")
    ) as HTMLElement[];
}

export function getCurrentOptionValues(options: HTMLElement[]) {
    const values: (string | null)[] = [];
    for (const opt of options) {
        values.push(opt.textContent);
    }
    return values;
}

export function getClearButton(coordinateSearchGroup: Element) {
    const buttonDiv = coordinateSearchGroup.querySelector(".chakra-input__right-element");
    if (!buttonDiv) {
        throw new Error("buttons not rendered");
    }
    const clearButton = buttonDiv.querySelector(".coordinate-input-clear-button");
    if (!clearButton) {
        throw new Error("clear button not rendered");
    }
    return clearButton;
}

export function getCopyButton(coordinateSearchGroup: Element) {
    const buttonDiv = coordinateSearchGroup.querySelector(".chakra-input__right-element");
    if (!buttonDiv) {
        throw new Error("buttons not rendered");
    }
    const copyButton = buttonDiv.querySelector(".coordinate-input-copy-button");
    if (!copyButton) {
        throw new Error("copy button not rendered");
    }
    return copyButton;
}
