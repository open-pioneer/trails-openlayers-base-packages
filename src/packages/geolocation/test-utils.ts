// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import OlMap from "ol/Map";
import { Circle, Fill, Stroke, Style } from "ol/style";
import { vi } from "vitest";
import { GeolocationController } from "./GeolocationController";

export function mockSuccessGeolocation(coords: number[]) {
    vi.spyOn(navigator, "geolocation", "get").mockReturnValue({
        clearWatch() {},
        getCurrentPosition() {},
        watchPosition(success) {
            setTimeout(() => {
                success({
                    coords: {
                        latitude: coords[0],
                        longitude: coords[1],
                        accuracy: 2500,
                        altitude: null,
                        altitudeAccuracy: null,
                        heading: null,
                        speed: null
                    }
                } as GeolocationPosition);
            }, 1);

            return 123;
        }
    } satisfies Partial<Geolocation>);
}

export function mockErrorGeolocation() {
    vi.spyOn(navigator, "geolocation", "get").mockReturnValue({
        clearWatch() {},
        getCurrentPosition() {},
        watchPosition(success, error) {
            setTimeout(() => {
                error?.({
                    code: 2,
                    message: "POSITION_UNAVAILABLE",
                    PERMISSION_DENIED: 1,
                    POSITION_UNAVAILABLE: 2,
                    TIMEOUT: 3
                });
            }, 1);

            return 123;
        }
    } satisfies Partial<Geolocation>);
}

export function setup() {
    const olMap = new OlMap();
    return {
        olMap,
        controller: new GeolocationController(olMap, (error) => {
            console.error("Unexpected error", error);
        })
    };
}

export function setupWithCustomProperties() {
    const olMap = new OlMap();
    const maxZoom: number = getCustomMaxZoom();
    const positionFeatureStyle: Style = getCustomPositionStyle();
    const accuracyFeatureStyle: Style = getCustomAccuracyStyle();
    const trackingOptions: PositionOptions = getCustomTrackingOptions();

    const controller = new GeolocationController(
        olMap,
        (error) => {
            console.error("Unexpected error", error);
        },
        trackingOptions
    );
    controller.setMaxZoom(maxZoom);
    controller.setPositionFeatureStyle(positionFeatureStyle);
    controller.setAccuracyFeatureStyle(accuracyFeatureStyle);
    return {
        olMap,
        controller
    };
}

export function getCustomMaxZoom() {
    return 20;
}
export function getCustomPositionStyle() {
    return new Style({
        image: new Circle({
            fill: new Fill({
                color: "rgba(255,255,255,0.4)"
            }),
            stroke: new Stroke({
                color: "#3399CC",
                width: 1.25
            }),
            radius: 5
        }),
        fill: new Fill({
            color: "rgba(255,255,255,0.4)"
        }),
        stroke: new Stroke({
            color: "#3399CC",
            width: 1.25
        })
    });
}

export function getCustomAccuracyStyle() {
    return new Style({
        image: new Circle({
            fill: new Fill({
                color: "rgba(255,255,255,0.4)"
            }),
            stroke: new Stroke({
                color: "#3399CC",
                width: 1.25
            }),
            radius: 5
        }),
        fill: new Fill({
            color: "rgba(255,255,255,0.4)"
        }),
        stroke: new Stroke({
            color: "#3399CC",
            width: 1.25
        })
    });
}

function getCustomTrackingOptions() {
    return {
        "enableHighAccuracy": true,
        "timeout": 20,
        "maximumAge": 50
    };
}
