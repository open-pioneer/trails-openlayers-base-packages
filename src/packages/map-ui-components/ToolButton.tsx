// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { Button, ButtonProps, Toggle} from "@chakra-ui/react";
import { Tooltip, TooltipProps} from "@open-pioneer/chakra-snippets/tooltip";
import {
    FC,
    MouseEvent,
    MouseEventHandler,
    ReactElement,
    RefAttributes,
    forwardRef,
    useState
} from "react";
import { CommonComponentProps, useCommonComponentProps } from "@open-pioneer/react-utils";
import classNames from "classnames";

/**
 * Properties supported by {@link ToolButton}.
 */
export interface ToolButtonProps extends CommonComponentProps, RefAttributes<HTMLButtonElement> {
    /**
     * The label for the ToolButton.
     * This value services as the tooltip text and the aria-label.
     *
     * This property is required for a11y reasons because a ToolButton usually only displays an icon.
     */
    label: string;

    /**
     * The icon displayed by the button.
     */
    icon: ReactElement;

    /**
     * If `true`, the button will show a spinner.
     * Defaults to `false`.
     */
    loading?: boolean;

    /**
     * If `true`, indicates that the button is currently active with a different style.
     * Defaults to `undefined`.
     *
     * A value of `true` or `false` indicates that the button supports being active (i.e. pressed).
     * In that case the `aria-pressed` attribute will be configured automatically
     * (see https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-pressed).
     */
    active?: boolean;

    /**
     * If `true`, the button will be disabled.
     * Defaults to `false`.
     */
    disabled?: boolean;

    /**
     * The callback that will be called when the user clicks the button.
     */
    onClick?: MouseEventHandler<HTMLButtonElement> | undefined;

    /**
     * Additional properties for the `Tooltip` element.
     *
     * Note that the ToolButton also defines some of these props.
     */
    tooltipProps?: Partial<TooltipProps>;

    /**
     * Additional properties for the `Button` element.
     *
     * Note that the ToolButton also defines some of these props.
     */
    buttonProps?: Partial<ButtonProps>;
}

/**
 * An button with a tooltip, used for tool buttons on a map.
 */
export const ToolButton: FC<ToolButtonProps> = forwardRef(function ToolButton(
    props: ToolButtonProps,
    ref
) {
    const {
        label,
        icon,
        onClick: onClickProp,
        loading,
        disabled,
        active,
        tooltipProps,
        buttonProps
    } = props;

    const {
        containerProps: { className: baseClassName, ...containerProps }
    } = useCommonComponentProps("tool-button", props);

    const className = classNames(baseClassName, {
        "tool-button--active": active,
        "tool-button--loading": loading,
        "tool-button--disabled": disabled
    });

    const [tooltipOpen, setTooltipOpen] = useState(false);
    const onClick = (e: MouseEvent<HTMLButtonElement>) => {
        // Immediately hide the tooltip. When the label switches in reaction to the click,
        // the tooltip would flicker briefly with the new label.
        setTooltipOpen(false);
        onClickProp?.(e);
    };

    return (
        <Tooltip
            content={label}
            openDelay={500}
            {...tooltipProps}
            /* don't allow overwrite because component would break */
            open={tooltipOpen}
            onOpenChange={(e) => setTooltipOpen(e.open)}
        >
            <Toggle.Root pressed={active} /*toggles aria-pressed*/ asChild>
                <ButtonIgnoringAriaProps
                    className={className}
                    ref={ref}
                    aria-label={label}
                    padding={0}
                    disabled={disabled}
                    loading={loading}
                    {...containerProps}
                    {...buttonProps}
                    /* don't allow overwrite because component would break */
                    onClick={onClick}
                >
                    {icon}
                </ButtonIgnoringAriaProps>
            </Toggle.Root>
        </Tooltip>
    );
});

/**
 * The tooltip will automatically set 'aria-describedby' when it is being shown.
 * This is redundant because the aria-label already has the same content as the tooltip.
 * This component wraps chakra's button to ignore the *-by attributes.
 */
const ButtonIgnoringAriaProps = forwardRef(function ButtonIgnoringAriaProps(
    props: ButtonProps,
    ref: React.ForwardedRef<HTMLButtonElement>
) {
    const { "aria-labelledby": _label, "aria-describedby": _describedBy, ...rest } = props;
    return <Button ref={ref} {...rest} />;
});
