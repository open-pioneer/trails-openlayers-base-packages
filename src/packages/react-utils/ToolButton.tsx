// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { FC, MouseEventHandler, ReactElement, RefAttributes, forwardRef } from "react";
import { CommonComponentProps, useCommonComponentProps } from "./useCommonComponentProps";
import { Button, ButtonProps, Tooltip, TooltipProps } from "@open-pioneer/chakra-integration";

/**
 * Properties supported by {@link ToolButton}.
 */
export interface ToolButtonProps extends CommonComponentProps, RefAttributes<HTMLButtonElement> {
    /**
     * The label for the ToolButton.
     * This value services as the tooltip text and the aria-label.
     *
     * This property is required for a11y reasons because a ToolButton usually only displays an icon.
     */
    label: string;

    /**
     * The icon displayed by the button.
     */
    icon: ReactElement;

    /**
     * The callback that will be called when the user clicks the button.
     */
    onClick?: MouseEventHandler<HTMLButtonElement> | undefined;

    /**
     * Additional properties for the `Tooltip` element.
     *
     * Note that the ToolButton also defines some of these props.
     */
    tooltipProps?: Partial<TooltipProps>;

    /**
     * Additional properties for the `Button` element.
     *
     * Note that the ToolButton also defines some of these props.
     */
    buttonProps?: Partial<ButtonProps>;
}

/**
 * An button with a tooltip, used for tool buttons on a map.
 */
export const ToolButton: FC<ToolButtonProps> = forwardRef(function ToolButton(
    props: ToolButtonProps,
    ref
) {
    const { label, icon, onClick, tooltipProps, buttonProps } = props;
    const { containerProps } = useCommonComponentProps("tool-button", props);

    return (
        <Tooltip label={label} placement="auto" openDelay={500} {...tooltipProps}>
            <ButtonIgnoringAriaProps
                ref={ref}
                aria-label={label}
                leftIcon={icon}
                onClick={onClick}
                iconSpacing={0}
                padding={0}
                {...containerProps}
                {...buttonProps}
            />
        </Tooltip>
    );
});

/**
 * The tooltip will automatically set 'aria-describedby' when it is being shown.
 * This is redundant because the aria-label already has the same content as the tooltip.
 * This component wraps chakra's button to ignore the *-by attributes.
 */
const ButtonIgnoringAriaProps = forwardRef(function ButtonIgnoringAriaProps(
    props: ButtonProps,
    ref: React.ForwardedRef<HTMLButtonElement>
) {
    const { "aria-labelledby": _label, "aria-describedby": _describedBy, ...rest } = props;
    return <Button ref={ref} {...rest} />;
});
