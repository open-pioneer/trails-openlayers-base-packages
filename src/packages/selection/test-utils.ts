// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { Reactive, reactive } from "@conterra/reactivity-core";
import {
    SelectionSource,
    SelectionResult,
    SelectionOptions,
    SelectionSourceStatus,
    SelectionKind
} from "./api";
import { Point } from "ol/geom";

export const fakeSelectedPointFeatures = [
    new Point([407354, 5754673]), // con terra (Bottom Right)
    new Point([404740, 5757893]) // Schloss (Top Left)
];

let count = 0;

export class FakePointSelectionSource implements SelectionSource {
    readonly label = `Fake Selection Source #${++count}`;
    #timeout: number;
    #status: Reactive<SelectionSourceStatus>;
    #pointFeatures: Point[];

    constructor(
        timeout?: number,
        status?: SelectionSourceStatus,
        pointFeatures: Point[] = fakeSelectedPointFeatures
    ) {
        this.#timeout = timeout || 0;
        this.#status = reactive(status || "unavailable");
        this.#pointFeatures = pointFeatures;
    }

    get status(): SelectionSourceStatus {
        return this.#status.value;
    }

    set status(value: SelectionSourceStatus) {
        this.#status.value = value;
    }

    async select(
        selectionKind: SelectionKind,
        options: SelectionOptions
    ): Promise<SelectionResult[]> {
        if (selectionKind.type !== "extent") {
            throw new Error(`Unsupported selection kind: ${selectionKind.type}`);
        }

        if (this.#status.value !== "available") return [];

        await new Promise((resolve) => setTimeout(resolve, this.#timeout));

        const allPoints = this.#pointFeatures.map((point, index) => {
            const result: SelectionResult = {
                id: index,
                geometry: point
            };
            if (!point.intersectsExtent(selectionKind.extent)) {
                return undefined;
            }
            return result;
        });

        const selectedPoints = allPoints.filter((s): s is SelectionResult => s != null);
        const limitedResults =
            selectedPoints.length > options.maxResults
                ? selectedPoints.slice(0, options.maxResults)
                : selectedPoints;
        return limitedResults;
    }
}

/**
 * For testing purposes only
 */
export class NoStatusSelectionSource implements SelectionSource {
    readonly label: string = "Testlabel";

    async select(_: SelectionKind, __: SelectionOptions): Promise<SelectionResult[]> {
        const allPoints = fakeSelectedPointFeatures.map((point, index) => {
            const result: SelectionResult = {
                id: index,
                geometry: point
            };
            return result;
        });
        return allPoints;
    }
}
