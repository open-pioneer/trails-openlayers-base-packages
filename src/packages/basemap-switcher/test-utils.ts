// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { MapConfig, MapConfigProvider, MapRegistry } from "@open-pioneer/map";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";
import Stamen from "ol/source/Stamen";
import { createService } from "@open-pioneer/test-utils/services";
import { MapRegistryImpl } from "@open-pioneer/map/MapRegistryImpl";
import { PackageContextProviderProps } from "@open-pioneer/test-utils/react";
import { screen, waitFor } from "@testing-library/react";

// used to avoid a "ResizeObserver is not defined" error
import ResizeObserver from "resize-observer-polyfill";
global.ResizeObserver = ResizeObserver;

export interface SimpleMapOptions {
    center?: { x: number; y: number };
    zoom?: number;
}

export const MAP_ID = "test";

export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}
export async function setupMap(options?: SimpleMapOptions) {
    const mapId = "test";
    const mapConfig: MapConfig = {
        initialView: {
            kind: "position",
            center: options?.center ?? { x: 847541, y: 6793584 },
            zoom: options?.zoom ?? 10
        },
        projection: "EPSG:3857",
        layers: [
            {
                id: "b-1",
                title: "OSM",
                isBaseLayer: true,
                visible: false,
                layer: new TileLayer({
                    source: new OSM()
                })
            },
            {
                id: "b-2",
                title: "Toner",
                isBaseLayer: true,
                visible: true,
                layer: new TileLayer({
                    source: new Stamen({ layer: "toner" })
                })
            }
        ]
    };

    const registry = await createService(MapRegistryImpl, {
        references: {
            providers: [new MapConfigProviderImpl(mapId, mapConfig)]
        }
    });

    return { mapId, registry };
}

export function createPackageContextProviderProps(
    service: MapRegistry
): PackageContextProviderProps {
    return {
        services: {
            "map.MapRegistry": service
        }
    };
}

class MapConfigProviderImpl implements MapConfigProvider {
    mapId = "default";
    mapConfig: MapConfig;

    constructor(mapId: string, mapConfig?: MapConfig | undefined) {
        this.mapId = mapId;
        this.mapConfig = mapConfig ?? {};
    }

    getMapConfig(): Promise<MapConfig> {
        return Promise.resolve(this.mapConfig);
    }
}
