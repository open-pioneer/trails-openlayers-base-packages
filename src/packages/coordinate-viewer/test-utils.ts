// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { OlMapConfigurationProvider } from "@open-pioneer/experimental-ol-map";
import { OlMapRegistry } from "@open-pioneer/experimental-ol-map/services";
import { ServiceOptions } from "@open-pioneer/runtime";
import { PackageContextProviderProps } from "@open-pioneer/test-utils/react";
import { createService } from "@open-pioneer/test-utils/services";
import { waitFor, screen } from "@testing-library/react";
import { MapOptions } from "ol/Map";
import View from "ol/View";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";

// used to avoid a "ResizeObserver is not defined" error
import ResizeObserver from "resize-observer-polyfill";
global.ResizeObserver = ResizeObserver;

export interface SimpleMapOptions {
    center?: [number, number];
    zoom?: number;
}

export const MAP_ID = "test";

export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}

export async function setupMap(options?: SimpleMapOptions) {
    const mapOptions: MapOptions = {
        view: new View({
            projection: "EPSG:3857",
            center: options?.center ?? [847541, 6793584],
            zoom: options?.zoom ?? 10
        }),
        layers: [
            new TileLayer({
                source: new OSM(),
                properties: { title: "OSM" }
            })
        ]
    };

    const mapConfigProvider = await createService(MapConfigProvider, {
        properties: {
            mapOptions: mapOptions,
            mapId: MAP_ID
        }
    });
    const registry = await createService(OlMapRegistry, {
        references: {
            providers: [mapConfigProvider]
        }
    });

    return { mapId: MAP_ID, registry };
}

export function createPackageContextProviderProps(
    service: OlMapRegistry
): PackageContextProviderProps {
    return {
        services: {
            "ol-map.MapRegistry": service
        }
    };
}

class MapConfigProvider implements OlMapConfigurationProvider {
    mapId = "default";
    mapOptions: MapOptions = {};

    constructor(options: ServiceOptions) {
        if (options.properties.mapOptions) {
            this.mapOptions = options.properties.mapOptions as MapOptions;
        }
        if (options.properties.mapId) {
            this.mapId = options.properties.mapId as string;
        }
    }

    getMapOptions(): Promise<MapOptions> {
        return Promise.resolve(this.mapOptions);
    }
}
