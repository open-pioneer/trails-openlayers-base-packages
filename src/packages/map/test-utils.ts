// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { PackageContextProviderProps } from "@open-pioneer/test-utils/react";
import { createService } from "@open-pioneer/test-utils/services";
import { screen, waitFor } from "@testing-library/react";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";
import {
    ExtentConfig,
    InitialViewConfig,
    LayerConfig,
    MapConfig,
    MapConfigProvider,
    MapRegistry,
    OlMapOptions
} from "./api";
import { MapRegistryImpl } from "./services";

// used to avoid a "ResizeObserver is not defined" error
import ResizeObserver from "resize-observer-polyfill";
global.ResizeObserver = ResizeObserver;

export interface SimpleMapOptions {
    center?: { x: number; y: number };
    zoom?: number;
    extent?: ExtentConfig;
    projection?: string;
    layers?: LayerConfig[];
    advanced?: OlMapOptions;

    noInitialView?: boolean;
    noProjection?: boolean;
}

export const MAP_ID = "test";

export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}

export async function setupMap(options?: SimpleMapOptions) {
    const mapId = "test";
    const getInitialView = (): InitialViewConfig => {
        if (options?.extent) {
            return {
                kind: "extent",
                extent: options.extent
            };
        }
        return {
            kind: "position",
            center: options?.center ?? { x: 847541, y: 6793584 },
            zoom: options?.zoom ?? 10
        };
    };

    const mapConfig: MapConfig = {
        initialView: options?.noInitialView ? undefined : getInitialView(),
        projection: options?.noProjection ? undefined : options?.projection ?? "EPSG:3857",
        layers: options?.layers ?? [
            {
                title: "OSM",
                layer: new TileLayer({
                    source: new OSM()
                })
            }
        ],
        advanced: options?.advanced
    };

    const registry = await createService(MapRegistryImpl, {
        references: {
            providers: [new MapConfigProviderImpl(mapId, mapConfig)]
        }
    });

    return { mapId, registry };
}

export function createPackageContextProviderProps(
    service: MapRegistry
): PackageContextProviderProps {
    return {
        services: {
            "map.MapRegistry": service
        }
    };
}

class MapConfigProviderImpl implements MapConfigProvider {
    mapId = "default";
    mapConfig: MapConfig;

    constructor(mapId: string, mapConfig?: MapConfig | undefined) {
        this.mapId = mapId;
        this.mapConfig = mapConfig ?? {};
    }

    getMapConfig(): Promise<MapConfig> {
        return Promise.resolve(this.mapConfig);
    }
}
