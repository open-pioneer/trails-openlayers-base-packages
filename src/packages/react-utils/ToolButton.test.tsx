// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { PackageContextProvider } from "@open-pioneer/test-utils/react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { expect, it } from "vitest";
import { ToolButton } from "./ToolButton";
import { createRef } from "react";

it("renders as a button", async () => {
    render(
        <PackageContextProvider>
            <ToolButton label="Button Label" icon={getIcon()} />
        </PackageContextProvider>
    );

    const button = screen.getByLabelText("Button Label");
    expect(button.tagName).toBe("BUTTON");
    expect(button).toMatchSnapshot();
});

it("calls event handler when clicked", async () => {
    const user = userEvent.setup();

    let clicked = 0;
    render(
        <PackageContextProvider>
            <ToolButton label="Button Label" icon={getIcon()} onClick={() => ++clicked} />
        </PackageContextProvider>
    );

    const button = screen.getByLabelText("Button Label");
    await user.click(button);
    expect(clicked).toBe(1);
});

it("allows configuration of additional button props", async () => {
    render(
        <PackageContextProvider>
            <ToolButton
                label="Button Label"
                icon={getIcon()}
                buttonProps={{
                    title: "foo"
                }}
            />
        </PackageContextProvider>
    );

    const button = screen.getByLabelText("Button Label");
    expect(button.title).toBe("foo");
});

it("automatically sets the 'aria-pressed' attribute when active", async () => {
    const getUI = (isActive: boolean | undefined) => (
        <PackageContextProvider>
            <ToolButton label="Button Label" icon={getIcon()} isActive={isActive} />
        </PackageContextProvider>
    );

    const { rerender } = render(getUI(false));

    // isActive=false -> aria-pressed=false
    const button = screen.getByLabelText("Button Label");
    expect(button.getAttribute("aria-pressed")).toBe("false");

    // isActive=true -> aria-pressed=true
    rerender(getUI(true));
    expect(button.getAttribute("aria-pressed")).toBe("true");

    // isActive=undefined -> aria-pressed not set
    rerender(getUI(undefined));
    expect(button.getAttribute("aria-pressed")).toBe(null);
});

it("shows a tooltip when hovered", async () => {
    const user = userEvent.setup();
    render(
        <PackageContextProvider>
            <ToolButton
                label="Button Label"
                icon={getIcon()}
                tooltipProps={{
                    openDelay: 0
                }}
            />
        </PackageContextProvider>
    );

    const button = screen.getByLabelText("Button Label");
    await user.hover(button);

    const tooltips = screen.getAllByRole("tooltip");
    expect(tooltips).toHaveLength(1);

    const tooltip = tooltips[0]!;
    expect(tooltip.textContent).toEqual("Button Label");

    // Does NOT set aria-describedby on the button (that would be the Tooltip's default behavior)
    // because we already have an aria-label with the same content.
    expect(button.getAttribute("aria-describedby")).toBe(null);
    expect(button.getAttribute("aria-labelledby")).toBe(null);
    expect(button.getAttribute("aria-label")).toBe("Button Label");
});

it("makes the button element available via 'ref'", async () => {
    const ref = createRef<HTMLButtonElement>();
    render(
        <PackageContextProvider>
            <ToolButton
                ref={ref}
                label="Button Label"
                icon={getIcon()}
                tooltipProps={{
                    openDelay: 0
                }}
            />
        </PackageContextProvider>
    );

    const button = screen.getByLabelText("Button Label");
    expect(ref.current).toBe(button);
});

function getIcon() {
    return <span>X</span>;
}
