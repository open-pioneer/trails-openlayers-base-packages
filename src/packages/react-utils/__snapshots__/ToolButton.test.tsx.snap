// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders as a button 1`] = `
<button
  aria-label="Button Label"
  class="chakra-button tool-button css-22xpxh"
  data-theme="light"
  type="button"
>
  <span
    class="chakra-button__icon css-1dwr89y"
    data-theme="light"
  >
    <span
      aria-hidden="true"
      focusable="false"
    >
      X
    </span>
  </span>
</button>
`;
